### ISSUE TITLE
Multiple m:n fields in one form

### ISSUE THREAD

**Issue by lupoalberto** [mootensai/yii2-relation-trait#17]
Hi again 😄 

it seems to me, that there is an issue if we use more then one m:n fields in the same form. 

> if (!$relSave || !empty($relModel->errors)) {
>                                         $relModelWords = 
> Inflector::camel2words(StringHelper::basename($AQ->modelClass));
>                                         $index++;
>                                         foreach ($relModel->errors as $validation) {
>                                             foreach ($validation as $errorMsg) {
>                                                 $this->addError($name, "$relModelWords #$index : $errorMsg");
>                                             }
>                                         }
>                                         $error = true;
>                                     } else {
>                                         //GET PK OF REL MODEL
>                                         if ($isManyMany) {
>                                             foreach ($relModel->primaryKey as $attr => $value) {
>                                                 $notDeletedPK[$i][$attr] = "'$value'";
>                                                 $fields[$attr] = "";

The problem is the $fields[$attr] Array. You never free it when a new relation is called and so all the attributes of the "old" relation remain also attributes of the new relation, even if they do not exist. I fixed this by setting $fields = [] at the beginning (in my case line 95 of your code). 
Now I'm able to save new data with more than one m:n relations, but deleting of items doesn't work anymore, if the items list is empty (works if there is at least one item left). What do you think?

**Comment by mootensai** [comment]
Could you give me your schema / mysql dump?? I've tried on sakila db and it's fine..

I'm currently doing this relation-trait by the way..

**Comment by lupoalberto** [comment]
Sorry, my english is not at best... However as you can see I attached a test database. There are 5 tables:
persons, country, countries, language, languages and several relations. Basically the countries and languages tables are the m:n raltion tables.

Now...everything works fine, as you can see, I created some persons with related data as countries and languages. The problem is, that if I try to update a person and remove for example all the related countries, than try to save it, nothing happens. I mean...everything seems to be ok, no error message, redirecting to the view page, etc. but when I reload the page all the countries are still there. If I remove only one item, everything is fine, the problem occurs only if I remove all the items of one related table. 

Also, everything works fine when I remove all the items of both related tables and even if one of the realted tables is already empty. However you can reproduce the issue by creating a person in my attached example. Then relate two countries and two languages. Next, save it. Everything should work. Now try to remove both related countries for example and try to save. As you can see, saving seems to be successfully but the removed countries are still there (no error message).

All the models, controller and views I created with IO Generator of gii

About the "fields" array mentioned in my first post here: in my opinion this should be automatically fixed at the moment when you implement the "composite primary key" as discussed in my other post, so don't care about that.
Thank you for your support and your great work!

[test.sql.zip](https://github.com/mootensai/yii2-relation-trait/files/308919/test.sql.zip)

**Comment by mootensai** [comment]
Thanks!

**Comment by lupoalberto** [comment]
Hello mootensai

I unfortunately think I found another mistake. In the attached example I have 4 tables: main (the main table), sub1 (first n:m relation), sub2 (second n:m relation), persons (the target table of the n:m relations).
All the models, controller and views I created with IO Generator of gii and leaved them untouched.

Now if you try to go the main view in yii2 everything works fine, as long as I put in values for only one subquery. At the moment I try to put in values for subquery1 and subquery2 it fails with this error message:

> SQLSTATE[21000]: Cardinality violation: 1241 Operand should contain 3 column(s)
The SQL being executed was: DELETE FROM `sub2` WHERE (`mainid`=1) AND ((mainid, personsid, somevalue) NOT IN (('1','1'), ('1','2'), ('1','3'), ('1','2','23')))

[example.sql.zip](https://github.com/mootensai/yii2-relation-trait/files/432491/example.sql.zip)

Greez
